ts/src/application.ts

```diff
@@ -26,30 +27,34 @@ class TodoSequence extends Sequence {
     try {
       // 1. Find the endpoint/route
       // Throws HttpErrors.NotFound when there is no route matching the request method and path
-      const {controllerName, methodName, spec, pathParams} = findRoute(request);
+      const {controllerName, methodName, spec, pathParams} = this.findRoute(request);
       
-      // 2. Authenticate & authorize
+      // 2. Authenticate and bind current user
       const user = await this.authenticate(request);
-      
-      // 3. Authorize the operation
-      // TODO(bajtos) how to pass target model id for owner-based ACLs?
-      await this.authorize(controllerName, methodName, user);
+
+      // Anything produced / determined during the sequence
+      // should be bound into the request context
+      this.context.bind('authentication.user').to(user);
       
       // 4. Parse (and validate!) the arguments
       // Throws HttpErrors.UnprocessableEntity when some of 
       // the parameters are not valid.
       const args = parseArgs(spec, request, pathParams);
+
+      // 3. Authorize the user should be able to invoke the operation
+      // These access controls operate on controller names,
+      // method names, user obj, and args
+      await this.authorize(controllerName, methodName, user, args);
       
       // 5. Invoke the controller method
-      // TODO(bajtos) how to pass the current user?
-      // Especially when this Sequence.run() wants to change
-      // the current user, in which case any bindings made by 
-      // this.authenticate() are bound to a wrong value.
-      // I would do the following:
-      //    const result = await this.invoke(controllerName, methodName, args, user);
       const result = await this.invoke(controllerName, methodName, args);
-      
-      // 6. Serialize the result (typically to JSON) and write HTTP response data
+
+      // 6. Authorize the result should be sent to the user
+      // These access controls operate on normal access meta as well
+      // as the result
+      await this.authorizeResult(controllerName, methodName, user, args, result);
```

Comment:
I don't fully understand this. What kind of authorization do you envision to happen on the result?

The first thing that comes to my mind is to filter out some data from the response depending on what the user is authorized to see. If that's the case, then I am concerned about performance implications of fetching unnecessary data from backend services, and also about security implications - e.g. if `authorizeResult` is implemented via a black-list of what should be removed, then such implementation can easily miss new data/properties/records that are added by future changes of the application.